### DISCUSSION TITLE
Spring과 Spring Boot의 차이

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#183] [Spring Boot]
.

**Comment by Kimprodp** [comment]
Spring은 Spring Framework의 핵심 모듈들을 기반으로 한 프레임워크로 엔터프라이즈 애플리케이션 개발을 지원하기 위한 대규모 오픈 소스 프로젝트이다. Spring Framework를 사용하기 위해서는 설정 파일 작성을 통한 스프링 컨테이너 구성, 필요한 빈 객체 등록 및 의존성 설정, 데이터베이스 연결, 트랜잭션 관리 등 다양한 설정을 개발자가 직접 수동으로 구성해야 했다.
따라서 프로젝트 초기화 과정에서 많은 설정과 의존성을 추가하게 되며 프로젝트는 시작하는데 시간이 많이 걸린다.
또한 스프링을 통해 웹 애플리케이션을 구축하기 위해서는 별도의 WAS를 설치하고 설정해야 했다.

Spring Boot는 Spring Framework를 기반으로 한 서브 프레임워크로 Spring의 문제점을 해결해주고, 더 쉽고 빠르게 스프링 애플리케이션을 개발할 수 있도록 해주는 도구이다.
Spring Boot를 사용하면 Spring에서 제공하는 여러 기능들을 자동으로 설정하여 개발자가 보다 쉽게 사용할 수 있도록 해준다.

**Spring Boot의 주요 특징**
1. 자동 설정(Auto Configuration)
- Spring Boot는 애플리케이션의 설정을 자동으로 구성한다.
- @EnableAutoConfiguration, @SpringBootApplication 어노테이션을 통해 자동 설정을 활성화

2. 의존성 관리 간소화
- 특정 기능을 쉽게 추가할 수 있도록 여러 개의 라이브러리와 의존성을 하나의 패키지로 묶어 제공하는 starter 의존성 통합 모듈을 제공한다.
- 예: spring-boot-starter-web, spring-boot-starter-data-jpa, spring-boot-starter-security

3. 내장 서버
- Tomcat, Jetty, Undertow와 같은 내장 웹 서버를 제공하여, 애플리케이션을 독립 실행형 JAR 파일로 배포하고, 바로 실행할 수 있게 한다.
- 배포를 위해 War 파일을 생성해서 Tomcat에 배포할 필요 없으며, JAR 파일에는 모든 의존성 라이브러리가 포함되어 있어 외부 서버 없이도 애플리케이션을 실행할 수 있다.

**Comment by seunghye218** [comment]
### Spring Framework

Spring Framework는 다양한 엔터프라이즈 애플리케이션을 개발하기 위한 포괄적인 프레임워크입니다. 

- 주요 기능
1. 의존성 주입 (Dependency Injection): 애플리케이션 컴포넌트 간의 의존성을 자동으로 관리합니다.
2. AOP (Aspect-Oriented Programming): 횡단 관심사(Cross-cutting Concerns)를 모듈화하여 코드의 분리와 재사용성을 높입니다.
3. 데이터 액세스: JDBC, JPA, Hibernate 등 다양한 데이터 액세스 기술을 지원합니다.
4. 트랜잭션 관리: 선언적 트랜잭션 관리를 지원합니다.
5. 모듈성: 다양한 모듈로 구성되어 필요에 따라 선택적으로 사용 가능합니다.

### Spring Boot

Spring Boot는 Spring Framework 기반 애플리케이션을 더욱 쉽게 개발할 수 있도록 도와주는 확장된 프레임워크입니다. 주로 빠른 애플리케이션 설정과 실행, 기본 설정 제공을 목표로 합니다.

- 주요 기능
1. 자동 설정 (Auto-configuration): 애플리케이션 설정을 자동으로 구성하여 개발자가 설정에 신경 쓸 필요가 없도록 합니다.
2. 스타터 의존성 (Starter Dependencies): 여러 관련 의존성을 한 번에 추가할 수 있는 의존성 모음 제공.
3. 임베디드 서버 (Embedded Server): 톰캣, 제티, 언더토우와 같은 내장 서버를 통해 독립 실행형 애플리케이션을 쉽게 배포할 수 있습니다.
4. Spring Initializr: 웹 기반 인터페이스로 빠르게 프로젝트를 생성할 수 있습니다.
5. 프로덕션 준비 기능: 모니터링, 보안 설정 등 프로덕션 환경에서 필요한 다양한 기능을 기본 제공.
